### PR TITLE
OSS-Fuzz issues

### PR DESCRIPTION
**Pull request by xhanulik**
Fixing the latest problems found by OSS-Fuzz.

##### Checklist
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/libopensc/pkcs15.c [hunk 1]
```diff
@@ -2465,6 +2465,10 @@ sc_pkcs15_read_file(struct sc_pkcs15_card *p15card, const struct sc_path *in_pat
 			sc_pkcs15_cache_file(p15card, in_path, data, len);
 		}
 	}
+	if (len == 0) {
+		free(data);
+		data = NULL;
+	}
```
Comment (Jakuje) [inline src/libopensc/pkcs15.c:2471]: I would keep this condition in the above branch, as the `data` is only allocated there and we already have a condition for caching results in there

**Review by Jakuje — CHANGES_REQUESTED** [review]

File: src/libopensc/card-starcos.c [hunk 2]
```diff
@@ -1002,7 +1005,12 @@ static int starcos_select_file(sc_card_t *card,
 			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_SUCCESS);
 		}
 		else
-			return starcos_select_aid(card, pathbuf, pathlen, file_out);
+		{
+			r = starcos_select_aid(card, pathbuf, pathlen, file_out);
+			if (r != SC_SUCCESS && file_out)
+				sc_file_free(*file_out);
+			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, r);
+		}
```
[the remark this answers is not in the document]
Comment (xhanulik) [reply src/libopensc/card-starcos.c:1013]: The overriding of `file_out` was already fixed by beaffda, where subsequent `starcos_select_*()` frees `file_out` before overwriting if there was something already.

These particular changes address the situation, where for example on line 976 `starcos_select_*()` allocates `file_out` for the first time, then on line 991 `file_out` is passed to another `starcos_select_*()`, which in the case it fails does not change `file_out` (so there is still the allocated memory from line 976). When this happens, I believe the memory should be freed in `starcos_select_file()`.

[the remark this answers is not in the document]
Comment (dengert) [reply src/libopensc/card-starcos.c:1013]: What it should do is check if response contains the AID as per ISO 7816-4 or something it understands. The muscle applet would respond 90 00 to any unknown select_file.

File: src/libopensc/muscle.c [hunk 3]
```diff
@@ -181,6 +181,9 @@ int msc_partial_update_object(sc_card_t *card, msc_id objectId, int offset, cons
 	sc_apdu_t apdu;
 	int r;
 
+	if (MSC_MAX_APDU - 9 <= dataLength)
```
[the remark this answers is not in the document]
Comment (xhanulik) [reply src/libopensc/muscle.c]: alright, fixed

File: src/libopensc/pkcs15.c (hunk 1 above)
Comment (xhanulik) [reply src/libopensc/pkcs15.c]: moved to the branch above

File: src/libopensc/card-starcos.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/card-starcos.c:1017]: This looks much better than tweaking the results after the call.

**Review by Jakuje — APPROVED** [review]

**Comment by frankmorgner** [comment]
I'm not sure if 52e87dfeb41edc0b63d5041805987d6958558b95 is the best fix for https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=54508. For me, it makes absolutely sense to free `file_out` exactly when assigning a new value to it (as in the original implementation, though I'd use `sc_file_free` rather than `free`)

AFAICT, the problem stems from assigning `file_out` to something which is not the requested file (i.e. some file that has been on the path to the requested file). If I see this correctly, that this change could be sufficient to avoid the memory leak:
```diff
diff --git a/src/libopensc/card-starcos.c b/src/libopensc/card-starcos.c
index 713f57637..a6a8a38fd 100644
--- a/src/libopensc/card-starcos.c
+++ b/src/libopensc/card-starcos.c
@@ -782,7 +782,7 @@ static int starcos_select_aid(sc_card_t *card,
                file->id = 0x0000;
                file->magic = SC_FILE_MAGIC;
 
-               free(*file_out);
+               sc_file_free(*file_out);
                *file_out = file;
        }
        SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_SUCCESS);
@@ -899,7 +899,7 @@ static int starcos_select_fid(sc_card_t *card,
                        file->size = 0;
                        file->namelen = 0;
                        file->magic = SC_FILE_MAGIC;
-                       free(*file_out);
+                       sc_file_free(*file_out);
                        *file_out = file;
                } else {
                        /* ok, assume we have a EF */
@@ -920,7 +920,7 @@ static int starcos_select_fid(sc_card_t *card,
                                return r;
                        }
 
-                       free(*file_out);
+                       sc_file_free(*file_out);
                        *file_out = file;
                }
        }
@@ -973,7 +973,7 @@ static int starcos_select_file(sc_card_t *card,
                                || card->cache.current_path.type != SC_PATH_TYPE_DF_NAME
                                || card->cache.current_path.len != pathlen
                                || memcmp(card->cache.current_path.value, in_path->aid.value, in_path->aid.len) != 0 ) {
-                               r = starcos_select_aid(card, in_path->aid.value, in_path->aid.len, file_out);
+                               r = starcos_select_aid(card, in_path->aid.value, in_path->aid.len, NULL);
                                LOG_TEST_RET(card->ctx, r, "Could not select AID!");
                        }
 
@@ -1093,6 +1093,7 @@ static int starcos_select_file(sc_card_t *card,
                                        file->size = 0;
                                        file->namelen = 0;
                                        file->magic = SC_FILE_MAGIC;
+                                       sc_file_free(*file_out);
                                        *file_out = file;
                                }
                                /* nothing left to do */
```
Note that we could also remove the lines setting `SC_FILE_MAGIC` as this is already done in `sc_file_new()`

**Comment by frankmorgner** [comment]
> 

To make it more concrete, I think all that is required would be to change https://github.com/OpenSC/OpenSC/blob/396856843f887f1f524fe76cf91ec6027c9a9a5b/src/libopensc/card-starcos.c#L976 to use `NULL` instead of `file_out`

**Comment by xhanulik** [comment]
@frankmorgner Thanks for looking into this, the suggested change seems to be the best fix for the problem. I will update it.

**Review by Jakuje — APPROVED** [review]

**Comment by frankmorgner** [comment]
thank you

**Pull request merged**